File: huffman.py
import heapq
from heapq import heappop, heappush


def isLeaf(root):
	return root.left is None and root.right is None

class Node:
	def __init__(self, char, freq, left=None, right=None):
		self.char = char
		self.freq = freq
		self.left = left
		self.right = right

	#This overriding is later used by heapq for comparing Nodes
	#Without overriding you would get the following error:
	#TypeError: '<' not supported between instances of 'Node' and 'Node'
	def __lt__(self, other):
		return self.freq < other.freq



def encode(root, string, huffman_code):

	if root is None:
		return

	#this is for going bottom-up
	if isLeaf(root):
		huffman_code[root.char] = string if len(string) > 0 else "1"

	encode(root.left, string + "0", huffman_code)
	encode(root.right, string + "1", huffman_code)

def decode(root, index, string):
	if root is None:
		return index

	if isLeaf(root):
		print(root.char, end="")
		return index

	index = index + 1
	root = root.left if string[index] == "0" else root.right
	return decode(root, index, string)

def createHuffmanTree(string):

	if len(string) == 0:
		return

	#count frequency of characters
	frequency = {i: string.count(i) for i in set(string)}

	#priority queue with Nodes for character and frequency pairs
	priQ = [Node(k, v) for k, v in frequency.items()]
	heapq.heapify(priQ)

	while len(priQ) != 1:

		#Remove 2 nodes of lowest frequency
		left = heappop(priQ)
		right = heappop(priQ)

		#add the frequencies to create a new node
		total = left.freq + right.freq
		heappush(priQ, Node(None, total, left, right))

	root = priQ[0]
	huffmanCode = dict()

	encode(root, "", huffmanCode)

	print ("Huffman Codes: ", huffmanCode)
	print ("Original String: ", string)

	#this encodedString is to be given to decode!
	encodedString = ""
	for s in string:
		encodedString += huffmanCode.get(s)

	print ("Encoded String: ", encodedString)
	print ("Decoded String: ", end=" ")

	if isLeaf(root):
		#for repeat characters
		while root.freq > 0:
			print (root.char, end="")
			root.freq = root.freq - 1
	else:
		#since index gets incremented inside decode()
		index = -1
		while index < len(encodedString) - 1:
			index = decode(root, index, encodedString)
	print()

if __name__ == "__main__":
	string = "Huffman Code Test Works!"
	createHuffmanTree(string)

File: src/huffman.java
import java.util.Map;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.Comparator;

class Node {
    Character ch;
    Integer freq;
    Node left = null;
    Node right = null;

    Node(Character ch, Integer freq) {
        this.ch = ch;
        this.freq = freq;
    }

    public Node(Character ch, Integer freq, Node left, Node right) {
        this.ch = ch;
        this.freq = freq;
        this.left = left;
        this.right = right;
    }
}
public class huffman {

    public static void encode(Node root, String string, Map<Character, String> huffmanCode) {
        if (root == null) {
            return;
        }

        if (isLeaf(root)) {
            huffmanCode.put(root.ch, string.length() > 0 ? string : "1");
        }

        encode(root.left, string + '0', huffmanCode);
        encode(root.right, string + '1', huffmanCode);
    }

    public static int decode(Node root, int index, StringBuilder string) {
        if (root == null) {
            return index;
        }

        if (isLeaf(root)) {
            System.out.print(root.ch);
            return index;
        }

        index++;
        root = (string.charAt(index) == '0') ? root.left : root.right;
        index = decode(root, index, string);
        return index;
    }

    public static boolean isLeaf(Node root) {
        return root.left == null && root.right == null;
    }

    public static void createHuffmanTree(String string) {
        if (string == null || string.length() == 0) {
            return;
        }

        Map<Character, Integer> frequency = new HashMap<>();
        for (char c: string.toCharArray()) {
            frequency.put(c, frequency.getOrDefault(c, 0)+1);
        }

        System.out.println(frequency);

        PriorityQueue<Node> priQ;
        priQ = new PriorityQueue<>(Comparator.comparingInt(value -> value.freq));

        for (var v: frequency.entrySet()){
            priQ.add(new Node(v.getKey(), v.getValue()));
        }

        System.out.println(priQ);

        while (priQ.size() != 1) {
            Node left = priQ.poll();
            Node right = priQ.poll();

            int total = left.freq + right.freq;
            priQ.add(new Node(null, total, left, right));
        }

        Node root = priQ.peek();

        Map<Character, String> huffmanCode = new HashMap<>();
        encode(root, "", huffmanCode);

        System.out.println("Huffman Codes: " + huffmanCode);
        System.out.println("Original String: " + string);

        StringBuilder encodedString = new StringBuilder();
        for (char c: string.toCharArray()) {
            encodedString.append(huffmanCode.get(c));
        }
        System.out.println("Encoded String: " + encodedString);
        System.out.print("Decoded String: ");

        if (isLeaf((root))) {
            while (root.freq > 0) {
                System.out.print(root.ch);
                root.freq = root.freq - 1;
            }
        }
        else {
            int index = -1;
            while (index < encodedString.length() - 1) {
                index = decode(root, index, encodedString);
            }
        }

    }

    public static void main(String[] args) {
        String t = "Huffman Code Test Works!";
        createHuffmanTree(t);
    }

}
